extension/src/popup/views/SignBlob/index.tsx

```diff
@@ -1,85 +1,141 @@
-  isExperimentalModeEnabled: boolean;
-  isHardwareWallet: boolean;
-  isPasswordRequired: boolean;
-  publicKey: string;
-  rejectAndClose: () => void;
-  setIsDropdownOpen: (isRequired: boolean) => void;
-  setIsPasswordRequired: (isRequired: boolean) => void;
-  setStartedHwSign: (hasStarted: boolean) => void;
-}
-
-export const SignBlobBody = ({
-  publicKey,
-  allAccounts,
-  isDropdownOpen,
-  handleApprove,
-  isConfirming,
-  accountNotFound,
-  rejectAndClose,
-  currentAccount,
-  setIsDropdownOpen,
-  setIsPasswordRequired,
-  isPasswordRequired,
-  blob,
-  isExperimentalModeEnabled,
-  hwStatus,
-  isHardwareWallet,
-  setStartedHwSign,
-}: SignBlobBodyProps) => {
-  const dispatch: AppDispatch = useDispatch();
+import { LedgerSign } from "popup/components/hardwareConnect/LedgerSign";
+import { SlideupModal } from "popup/components/SlideupModal";
+
+import { VerifyAccount } from "popup/views/VerifyAccount";
+import { BlobToSign, parsedSearchParam } from "helpers/urls";
+import { truncatedPublicKey } from "helpers/stellar";
+
+import "./styles.scss";
+
+export const SignBlob = () => {
+  const location = useLocation();
   const { t } = useTranslation();
+  const blob = parsedSearchParam(location.search) as BlobToSign;
   const { accountToSign, domain, isDomainListedAllowed } = blob;
 
+  const dispatch: AppDispatch = useDispatch();
+  const isExperimentalModeEnabled = useSelector(
+    settingsExperimentalModeSelector,
+  );
+
+  const hardwareWalletType = useSelector(hardwareWalletTypeSelector);
+  const isHardwareWallet = !!hardwareWalletType;
+  const {
+    hardwareWalletData: { status: hwStatus },
+  } = useSelector(transactionSubmissionSelector);
+
+  const [startedHwSign, setStartedHwSign] = useState(false);
+  const [currentAccount, setCurrentAccount] = useState({} as Account);
+  const [isPasswordRequired, setIsPasswordRequired] = useState(false);
+  const [isConfirming, setIsConfirming] = useState(false);
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const [accountNotFound, setAccountNotFound] = useState(false);
+
+  const allAccounts = useSelector(allAccountsSelector);
+  const publicKey = useSelector(publicKeySelector);
+  const hasPrivateKey = useSelector(hasPrivateKeySelector);
+
+  // the public key the user had selected before starting this flow
+  const defaultPublicKey = useRef(publicKey);
+  const allAccountsMap = useRef({} as { [key: string]: Account });
+
+  const rejectAndClose = () => {
+    dispatch(rejectBlob());
+    window.close();
+  };
+
+  const handleApprove = (signAndClose: () => Promise<void>) => async () => {
+    setIsConfirming(true);
+
+    if (hasPrivateKey) {
+      await signAndClose();
+    } else {
+      setIsPasswordRequired(true);
+    }
+
+    setIsConfirming(false);
+  };
+
+  useEffect(() => {
+    if (startedHwSign && hwStatus === ShowOverlayStatus.IDLE) {
+      window.close();
+    }
+  }, [startedHwSign, hwStatus]);
+
+  useEffect(() => {
```

Comment:
I think something you did a good job of in the PR to add sign-blob was put this "auto-select account/set current account" logic in one place and pass it to different views. I think you should do the same here so we're not repeating this logic in all the 3 `sign-X` views

Reply:
Hmm yeah I chose the repetition because they no longer share a common ancestor where it would make sense to hold common logic. I think the solution now would have to be a wrapper component that does the shared stuff and passes it to a generic child prop. Let me give that a shot then.

Reply:
Might also work as a custom hook?

Really want to avoid the repetition bc the logic is a little hairy

Reply:
made a new hook that worked pretty nicely for this in 47ea4f923d0ef3128e286610f5b783a9688ce084